igor-web/src/test/groovy/com/netflix/spinnaker/igor/docker/DockerMonitorSpec.groovy

```diff
@@ -26,172 +26,199 @@ import com.netflix.spinnaker.igor.history.EchoService
+      dockerRegistryCache,
+      dockerRegistryAccounts,
+      Optional.of(echoService),
+      Optional.of(keelService),
+      dockerRegistryProperties,
+      Mock(TaskScheduler))
+      .postEvent(cachedImages, taggedImage, "imageId")
+
+    then:
+    echoServiceCallCount * echoService.postEvent({ DockerEvent event ->
+      assert event.content.tag == taggedImage.tag
+      assert event.content.account == taggedImage.account
+      assert event.content.registry == taggedImage.registry
+      assert event.content.repository == taggedImage.repository
+      assert event.content.digest == taggedImage.digest
+      return true
+    })
+
+    when: "should short circuit if `echoService` is not available"
+    createSubject().postEvent(["imageId"] as Set, taggedImage, "imageId")
+
+    then:
+    notThrown(NullPointerException)
+
+    where:
+    cachedImages    || echoServiceCallCount
+    null            || 0
+    [] as Set       || 0
+    ["job1"] as Set || 1
+
+  }
+
+  void 'should include decorated artifact in the payload'() {
+    given:
+    def taggedImage = new TaggedImage(
+      tag: "tag",
+      account: "account",
+      registry: "registry",
+      repository: "repository",
+      digest: "digest",
+      newtLabels: ["buildNumber": "111", "commitId": "ab12c3"]
+    )
+
+    when:
+    createSubject().postEvent(["job1"] as Set, taggedImage, "imageId")
+
+    then:
+    1 * echoService.postEvent({ DockerEvent event ->
+      assert event.artifact.version == taggedImage.tag
+      assert event.artifact.name == taggedImage.repository
+      assert event.artifact.type == "docker"
+      assert event.artifact.reference == "registry/repository:tag"
+      assert event.artifact.metadata.registry == taggedImage.registry
+      return true
+    })
+    1 * keelService.sendArtifactEvent({ Map event ->
+      def artifacts = event.payload.artifacts
+      assert artifacts.size() == 1
+      assert artifacts[0].name == "repository"
+      assert artifacts[0].type == "DOCKER"
+      assert artifacts[0].metadata.tag == "tag"
+      assert artifacts[0].metadata.buildNumber == "111"
+      assert artifacts[0].metadata.commitId == "ab12c3"
+      return true
+    })
+  }
+
+  void 'should include not include build and commit details if newtLables is missing from taggedImage'() {
+    given:
+    def taggedImageWithoutMetadata = new TaggedImage(
+      tag: "new-tag",
+      account: "account",
+      registry: "registry",
+      repository: "repository",
+      digest: "digest"
+    )
+
+    when:
+    createSubject().postEvent(["job1"] as Set, taggedImageWithoutMetadata, "imageId")
+
+    then:
+    1 * keelService.sendArtifactEvent({ Map event ->
+      def artifacts = event.payload.artifacts
+      assert artifacts.size() == 1
+      assert artifacts[0].name == "repository"
+      assert artifacts[0].type == "DOCKER"
+      assert artifacts[0].metadata.tag == "new-tag"
+      assert artifacts[0].metadata.buildNumber == null
```

Comment:
is there a way to assert the key isn't present?

Reply:
yep! changed to `containsKey`.